File: packages/zodiac-js/src/calculator/western-calculator.ts
/**
 * Western Zodiac calculator.
 *
 * Determines the Western Zodiac sign from a month and day input.
 * Uses standard date-range approach.
 */

import { getAllSigns } from '../western/signs';
import { validateDate } from '../validation/validator';
import { getLocaleText, getSupportedLocales } from '../localization/locale-manager';
import { getWesternIcon } from '../icons/icon-provider';
import type { ZodiacResult } from '../models/zodiac-result';

export interface WesternOptions {
  /** Optional locale code for localized output (e.g. 'zh-CN', 'fr', 'ar') */
  locale?: string;
  /**
   * Icon mode:
   * 'none' - No icon included
   * 'builtin' - Uses the built-in SVG icon
   * 'custom' - Retains the iconId for custom asset loading
   * Default is 'builtin'.
   */
  iconMode?: 'none' | 'builtin' | 'custom';
}

/**
 * Determines the Western Zodiac sign for a given month and day.
 *
 * @param month The month of birth (1-12)
 * @param day The day of birth (1-31)
 * @param options Optional configuration for locale and icons
 * @returns The resolved ZodiacResult
 * @throws ZodiacValidationError if month or day is invalid
 */
export function getWesternSign(month: number, day: number, options?: WesternOptions): ZodiacResult {
  validateDate(month, day);

  const signs = getAllSigns();
  let foundSign = signs[0]; // Default fallback

  // Find the matching sign
  for (const sign of signs) {
    const { startMonth, startDay, endMonth, endDay } = sign.dateRange;

    // A sign always spans exactly two months. 
    // This logic correctly handles Capricorn wrapping from 12 to 1.
    if (
      (month === startMonth && day >= startDay) ||
      (month === endMonth && day <= endDay)
    ) {
      foundSign = sign;
      break;
    }
  }

  // Construct the result
  const result: ZodiacResult = {
    sign: foundSign,
  };

  // 1. Localization
  if (options?.locale) {
    const locale = options.locale;
    (result as any).locale = getSupportedLocales().includes(locale) ? locale : 'en';
    (result as any).localizedName = getLocaleText(locale, `western.${foundSign.id}.name`) ?? foundSign.name;
    (result as any).localizedDescription = getLocaleText(locale, `western.${foundSign.id}.description`) ?? foundSign.description;
    
    // Attempt to load localized keywords, fallback to English
    const localizedKeywords = getLocaleText(locale, `western.${foundSign.id}.keywords`);
    if (Array.isArray(localizedKeywords)) {
      (result as any).localizedKeywords = localizedKeywords;
    } else {
      (result as any).localizedKeywords = foundSign.keywords;
    }
  }

  // 2. Icon Provider
  const iconMode = options?.iconMode ?? 'builtin';
  if (iconMode !== 'none') {
    (result as any).icon = getWesternIcon(foundSign.id, iconMode);
  }

  return result;
}

File: packages/zodiac-dart/lib/src/calculator/western_calculator.dart
import '../models/zodiac_result.dart';
import '../western/signs.dart';
import '../validation/validator.dart';
import '../localization/locale_manager.dart';
import '../icons/icon_provider.dart';

class WesternOptions {
  final String? locale;
  final String iconMode; // 'none', 'builtin', 'custom'

  const WesternOptions({
    this.locale,
    this.iconMode = 'builtin',
  });
}

/// Determines the Western Zodiac sign for a given month and day.
ZodiacResult getWesternSign(int month, int day, {WesternOptions? options}) {
  validateDate(month, day);

  var foundSign = westernSigns[0];

  for (var sign in westernSigns) {
    final startMonth = sign.dateRange.startMonth;
    final startDay = sign.dateRange.startDay;
    final endMonth = sign.dateRange.endMonth;
    final endDay = sign.dateRange.endDay;

    if ((month == startMonth && day >= startDay) ||
        (month == endMonth && day <= endDay)) {
      foundSign = sign;
      break;
    }
  }

  String? resolvedLocale;
  String? localizedName;
  String? localizedDescription;
  List<String>? localizedKeywords;

  if (options?.locale != null) {
    final locale = options!.locale!;
    resolvedLocale = getSupportedLocales().contains(locale) ? locale : 'en';

    localizedName =
        getLocaleText(locale, 'western.${foundSign.id}.name') as String? ??
            foundSign.name;
    localizedDescription =
        getLocaleText(locale, 'western.${foundSign.id}.description')
                as String? ??
            foundSign.description;

    final keywords = getLocaleText(locale, 'western.${foundSign.id}.keywords');
    if (keywords is List) {
      localizedKeywords = keywords.cast<String>();
    } else {
      localizedKeywords = foundSign.keywords;
    }
  }

  final iconMode = options?.iconMode ?? 'builtin';
  final icon =
      iconMode != 'none' ? getWesternIcon(foundSign.id, iconMode) : null;

  return ZodiacResult(
    sign: foundSign,
    locale: resolvedLocale,
    localizedName: localizedName,
    localizedDescription: localizedDescription,
    localizedKeywords: localizedKeywords,
    icon: icon,
  );
}
